src/app/CommandHandler.cpp

```diff
@@ -157,8 +161,34 @@ CHIP_ERROR CommandHandler::ProcessInvokeRequest(System::PacketBufferHandle && pa
     {
         err = CHIP_NO_ERROR;
     }
-    ReturnErrorOnFailure(err);
-    return invokeRequestMessage.ExitContainer();
+    VerifyOrReturnError(err == CHIP_NO_ERROR, Status::InvalidAction);
+    VerifyOrReturnError(invokeRequestMessage.ExitContainer() == CHIP_NO_ERROR, Status::InvalidAction);
+    return Status::Success;
+}
+
+CHIP_ERROR CommandHandler::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PayloadHeader & aPayloadHeader,
+                                             System::PacketBufferHandle && aPayload)
+{
+    ChipLogDetail(DataManagement, "CommandHandler: Unexpected message type %d", aPayloadHeader.GetMessageType());
+    return OnUnknownMsgType();
+}
+
+CHIP_ERROR CommandHandler::OnUnknownMsgType()
+{
+    VerifyOrReturnError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    CHIP_ERROR err = StatusResponse::Send(Status::InvalidAction, mpExchangeCtx, false /*aExpectResponse*/);
+    if (err != CHIP_NO_ERROR)
+    {
+        // We have to manually close the exchange, because we called
+        // WillSendMessage already.
+        mpExchangeCtx->Close();
+    }
+    // Null out the (now-closed) exchange, so that when we try to
+    // SendCommandResponse() later (when our holdoff count drops to 0) it
+    // just fails and we don't double-respond.
+    mpExchangeCtx = nullptr;
```

Comment:
Comments should be added here as to why you're not calling Close() on the handler object (because there is an active CommandHandler::Handle somewhere in the system for us to even arrive here, since in all normal command processing flows, the response is dispatched synchronously, and the EC closed synhcronously as well.

Reply:
commented, thanks